analyzer/evmtokens/evm_tokens.go

```diff
@@ -0,0 +1,219 @@
+	rows, err := m.target.Query(ctx, m.qf.RuntimeEVMTokensAnalysisStaleQuery(), limit)
+	if err != nil {
+		return nil, fmt.Errorf("querying discovered tokens: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var staleToken StaleToken
+		if err = rows.Scan(
+			&staleToken.Addr,
+			&staleToken.LastMutateRound,
+			&staleToken.Type,
+			&staleToken.AddrContextIdentifier,
+			&staleToken.AddrContextVersion,
+			&staleToken.AddrData,
+		); err != nil {
+			return nil, fmt.Errorf("scanning discovered token: %w", err)
+		}
+		staleTokens = append(staleTokens, &staleToken)
+	}
+	return staleTokens, nil
+}
+
+func (m Main) processBatch(ctx context.Context) error {
+	staleTokens, err := m.getStaleTokens(ctx, MaxDownloadBatch)
+	if err != nil {
+		return fmt.Errorf("getting discovered tokens: %w", err)
+	}
+	m.logger.Info("processing", "num_stale_tokens", len(staleTokens))
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, DownloadTimeout)
+	defer cancel()
+	group, groupCtx := errgroup.WithContext(ctxWithTimeout)
+
+	batch := &storage.QueryBatch{}
+
+	for _, st := range staleTokens {
+		// Redeclare `st` for unclobbered use within goroutine.
+		staleToken := st
+		group.Go(func() error {
+			m.logger.Info("downloading", "stale_token", staleToken)
+			// todo: assert that addr context is secp256k1
+			if staleToken.Type == nil {
+				tokenData, err := modules.EVMDownloadNewToken(
+					groupCtx,
+					m.logger,
+					m.source,
+					staleToken.LastMutateRound,
+					staleToken.AddrData,
+				)
+				if err != nil {
+					return fmt.Errorf("downloading new token %s: %w", staleToken.Addr, err)
+				}
+				batch.Queue(m.qf.RuntimeEVMTokenInsertQuery(),
+					staleToken.Addr,
+					modules.EVMTokenTypeERC20,
+					tokenData.Name,
+					tokenData.Symbol,
+					tokenData.Decimals,
+					tokenData.TotalSupply.String(),
+				)
+			} else {
+				tokenData, err := modules.EVMDownloadMutatedToken(
+					groupCtx,
+					m.logger,
+					m.source,
+					staleToken.LastMutateRound,
+					staleToken.AddrData,
+					*staleToken.Type,
+				)
+				if err != nil {
+					return fmt.Errorf("downloading mutated token %s: %w", staleToken.Addr, err)
+				}
+				batch.Queue(m.qf.RuntimeEVMTokenUpdateQuery(),
+					staleToken.Addr,
+					tokenData.TotalSupply.String(),
+				)
+			}
+			batch.Queue(m.qf.RuntimeEVMTokenAnalysisUpdateQuery(), staleToken.Addr, st.LastMutateRound)
+			return nil
+		})
+	}
+
+	if err := group.Wait(); err != nil {
+		return err
+	}
+
+	if err := m.target.SendBatch(ctx, batch); err != nil {
+		return fmt.Errorf("sending batch: %w", err)
+	}
+	return nil
+}
+
+func (m Main) Start() {
+	ctx := context.Background()
+
+	backoff, err := util.NewBackoff(
+		100*time.Millisecond,
+		6*time.Second,
+		// ^cap the timeout at the expected
+		// emerald round time
+	)
+	if err != nil {
+		m.logger.Error("error configuring indexer backoff policy",
+			"err", err,
+		)
+		return
+	}
+
+	for {
+		backoff.Wait()
+
+		if err := m.processBatch(ctx); err != nil {
```

Comment:
`processBatch` should return the number of updated tokens. If 0, that means we're all caught up, and we should also call `backoff.Failure()`.
Currently, once the analyzer catches up, it does a series of `backoff.Success()` and keeps processing 0 tokens at max speed.

Reply:
yeah I was thinking that too

Reply:
added